libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -319,6 +322,82 @@ void AP_UAVCAN::init(uint8_t driver_index, bool enable_filters)
     debug_uavcan(2, "UAVCAN: init done\n\r");
 }
 
+// send ESC telemetry messages over MAVLink
+void AP_UAVCAN::send_esc_telemetry_mavlink(uint8_t mav_chan)
+{
+    static const uint8_t MAV_ESC_GROUPS = 3;
+    static const uint8_t MAV_ESC_PER_GROUP = 4;
+
+    bool group[MAV_ESC_GROUPS] = {false};
+
+    // each ESC Mavlink message contain group of 4 ESC.
+    // find what groups have at least one ESC available data,
+    // and mark those groups for sending. 
+    for (uint8_t i = 0; i < MAV_ESC_GROUPS; i++) { 
+        for (uint8_t j = 0; j < MAV_ESC_PER_GROUP; j++) {
+            if (_escs_data[(i*MAV_ESC_PER_GROUP)+j].available) {
+                group[i] = true; 
+                break;
+            }
+        }
+    }
+
+    for (uint8_t i = 0; i < MAV_ESC_GROUPS; i++) {
+
+        // skip this group of ESCs if no data available to send
+        if (!group[i]) {
+            continue;
+        }
+
+        // arrays to hold output
+        uint8_t temperature[MAV_ESC_PER_GROUP] {};
+        uint16_t voltage[MAV_ESC_PER_GROUP] {};
+        uint16_t current[MAV_ESC_PER_GROUP] {};
+        uint16_t current_tot[MAV_ESC_PER_GROUP] {};
+        uint16_t rpm[MAV_ESC_PER_GROUP] {};
+        uint16_t count[MAV_ESC_PER_GROUP] {};
+
+        // fill in output arrays
+        for (uint8_t j = 0; j < MAV_ESC_PER_GROUP; j++) {
+            uint8_t esc_id = i * MAV_ESC_PER_GROUP + j;
+            
+            WITH_SEMAPHORE(_telem_sem);
+            
+            if (!_escs_data[esc_id].available) {
```

Comment:
This is making the assumption that esc_id is within the bounds of `_escs_data`.

A bounds check with a break would be good.

The same comment applies to determining which groups to output.

What I'm worried about is someone reducing the number of servos to reduce resource consumption, and access-past-end-of-array happening here.

Reply:
Sure. Because we are now verifying the index in more than one place, I'll add a function for that to avoid duplicate code